aequilibrae/project/network/link_types.py

```diff
@@ -0,0 +1,129 @@
+from sqlite3 import IntegrityError, Connection
+from aequilibrae.project.network.link_type import LinkType
+from aequilibrae import logger
+from aequilibrae.project.table_loader import TableLoader
+
+
+class LinkTypes:
+    """
+    Access to the API resources to manipulate the link_types table in the network
+
+    ::
+
+        from aequilibrae import Project
+
+        p = Project()
+        p.open('path/to/project/folder')
+
+        link_types = p.network.link_types
+
+        # We can get a dictionary of all modes in the model
+        all_link_types = modes.all_types()
+
+        #And do a bulk change and save it
+        for link_type_id, link_type_obj in all_link_types.items():
+            link_type_obj.beta = 1
+
+        # We can save changes for all link types in one go
+        all_link_types.save()
+
+        # or just get one link_type in specific
+        default_link_type = link_types.get('y')
+
+        # or just get it by name
+        default_link_type = link_types.get_by_name('default')
+
+        # We can change the description of the mode
+        default_link_type.description = 'My own new description'
+
+        # Let's say we are using alpha to store lane capacity during the night as 90% of the standard
+        default_link_type.alpha =0.9 * default_link_type.lane_capacity
+
+        # To save this mode we can simply
+        default_link_type.save()
+
+        # We can also create a completely new link_type and add to the model
+        new_type = new('a')
+        new_type.link_type_name = 'Arterial'  # Only ASCII letters and *_* allowed
+        # other fields are not mandatory
+
+        # We then save it to the database
+        new_type.save()
+
+        # we can even keep editing and save it directly once we have added it to the project
+        new_type.lanes = 3
+        new_type.lane_capacity = 1100
+        new_type.save()
+
+    """
+    __items = {}
+
+    def __init__(self, net):
+        self.__all_types = []
```

Comment:
so much caching, it can't be necessary everywhere

I'd try to avoid it as a design pattern except where you make it explicit... like the concept of a network that you load/publish.

Reply:
The idea is similar to Emme's Load/publish, so the caching is necessary. Should I bring the save() up to the Network object only, then?

Reply:
What I'm saying though is that you are using class level variables to do it. I wasn't really clear with my language though, so apologies for that. 

Consider this example below - the two print(a) lines should produce the same output right? I haven't changed a in between calling them - but... that's not what happens.

I would advise that you have a single instance which manages the storage of this and then pass that around instead of shoving it into class variables and tracking whether they have been instantiated or not. 

```python
class A(object):
    my_var = []

    def __init__(self, bar):
        self.bar = bar
        self.my_var.append(bar)

    def __repr__(self):
        return f"{self.bar} -> {self.my_var}"


a = A(9)
print(a)    # 9 -> [9]
b = A(7)
c = A(8)
print(a)    # 9 -> [9, 7, 8]
```

Reply:
That was intentional.  What I don't want is two sets of link types (or modes) floating around